packages/flutter/lib/src/material/text_selection.dart

```diff
@@ -38,30 +40,103 @@ class _TextSelectionToolbar extends StatelessWidget {
   final VoidCallback handlePaste;
   final VoidCallback handleSelectAll;
 
+  @override
+  __TextSelectionToolbarState createState() => __TextSelectionToolbarState();
+}
+
+class __TextSelectionToolbarState extends State<_TextSelectionToolbar> {
+  // effective when items > 3
+  bool flipToMenu = false;
+
   @override
   Widget build(BuildContext context) {
     final List<Widget> items = <Widget>[];
-    final MaterialLocalizations localizations = MaterialLocalizations.of(context);
-
-    if (handleCut != null)
-      items.add(FlatButton(child: Text(localizations.cutButtonLabel), onPressed: handleCut));
-    if (handleCopy != null)
-      items.add(FlatButton(child: Text(localizations.copyButtonLabel), onPressed: handleCopy));
-    if (handlePaste != null)
-      items.add(FlatButton(child: Text(localizations.pasteButtonLabel), onPressed: handlePaste,));
-    if (handleSelectAll != null)
-      items.add(FlatButton(child: Text(localizations.selectAllButtonLabel), onPressed: handleSelectAll));
+    final MaterialLocalizations localizations =
+        MaterialLocalizations.of(context);
+
+    if (widget.handleCut != null)
+      items.add(FlatButton(
+          child: Text(localizations.cutButtonLabel),
+          onPressed: widget.handleCut));
+    if (widget.handleCopy != null)
+      items.add(FlatButton(
+          child: Text(localizations.copyButtonLabel),
+          onPressed: widget.handleCopy));
+    if (widget.handlePaste != null)
+      items.add(FlatButton(
+        child: Text(localizations.pasteButtonLabel),
+        onPressed: widget.handlePaste,
+      ));
+    if (widget.handleSelectAll != null)
+      items.add(FlatButton(
+          child: Text(localizations.selectAllButtonLabel),
+          onPressed: widget.handleSelectAll));
 
     // If there is no option available, build an empty widget.
     if (items.isEmpty) {
       return Container(width: 0.0, height: 0.0);
     }
 
+    if (items.length <= 3) {
+      return _buildSelectionToolbar(items);
+    } else {
+      final List<Widget> displayItems = <Widget>[
+        items[0],
+        items[1],
+        items[2],
+        IconButton(
+          icon: const Icon(Icons.more_vert),
+          onPressed: () {
+            setState(() {
+              flipToMenu = true;
+            });
+          },
+        ),
+      ];
+
+      if (flipToMenu) {
+        return _buildSelectionMenu(localizations.selectAllButtonLabel);
+      } else {
+        return _buildSelectionToolbar(displayItems);
+      }
+    }
+  }
+
+  Widget _buildSelectionMenu(String selectAllButtonLabel) {
+    return Material(
```

Comment:
This is taller than the normal selection menu, so it can overflow the screen independently of it, and it needs to handle this problem independently.  In native, if you open the selection menu very near to the top of the screen but where it still renders above the text, then open the submenu, the submenu will know to render the list down instead of up.

Here is where the selection menu does this to get you started:

https://github.com/flutter/flutter/blob/d33cf1155678c04e858c23df62432c53e7869d3d/packages/flutter/lib/src/material/text_selection.dart#L163